### DISCUSSION TITLE
Can not connet robot with robot@10.42.0.1

### DISCUSSION THREAD

**Discussion by Lilly-yang** [prl-mushr/mushr#26] [Software]
Hi,

It's me again. :)

I am now following the [navigation](https://mushr.io/tutorials/navigation/) and [first step](https://mushr.io/tutorials/first_steps/) tutorials. I run into some problems when I tried to connect the robot to the rivz on my laptop.

- Q1: I used to connect the robot to wifi here in our library, and then use its IP to connect to my laptop instead of the static IP (robot@10.42.0.1), because it can't work properly. The dynamic IP is good, but it changes a lot. I'm sure I followed every step of [build instruction](https://mushr.io/hardware/build_instructions/) and checked it with 'ifconfig', its IP address is 10.42.0.1. Could you please give me any clues (hardware or software) why it not working properly?

- Q2: Following the [first step](https://mushr.io/tutorials/first_steps/) tutorial, I set ROS_IP=192.168.18.58 (my laptop IP) 
![image](https://user-images.githubusercontent.com/39558023/119978129-8e2b7600-bfb9-11eb-824f-b1b6bce91a54.png)
But for the ROS_MASTER_URI=http://10.42.0.1:11311. In addition to the disconnect problem mentioned in Q1. Could you please tell me how to get the information at the end? Or it is always '11311'?

Thank you!

**Comment by schmittlema** [comment]
Hi Lilly!
Great questions, we encounter these problems often internally and have been thinking about how to make the whole ROS_MASTER_URI and ROS_IP thing easier to follow.

Q1: So are you trying to have the car connect to the library wifi but keep a static IP? If so, at the bottom of the first steps tutorial there is a Wifi setup section that goes over how to make sure that IP address doesn't change. If you followed that and it still isn't working, it could be a problem stemming from how the library wifi is setup. To test if the problem is specific to your library wifi, you could try connecting the car to a home network or a hotspot following the same steps. Setting a static IP on those networks have worked for us in the past.

Q2: This is a super common problem. The [workflow/troubleshooting tutorial ](https://mushr.io/tutorials/workflow/)goes over this a bit, but I'll try to restate here a bit more clearly. When you run teleop on your robot and that is the first thing you launch it creates a ROS MASTER which is kinda the center for all communication. ROS_MASTER_URI tells the ROS system on the given computer (laptop or car) where to find this ROS MASTER. So on the car it should be its default, `http://localhost:11311` since the car created the ROS MASTER. On the laptop, you need to set the `ROS_MASTER_URI` to point to the car. So this should be http://CAR_IP:11311. Wher CAR_IP is the IP address of the car, obtained by using tools like `ifconfig`. 

`ROS_IP` is always set to the IP of the given computer. So you correctly set this to your laptop, you must also set it on the car to the car's IP.

11311 port is the default ROS port and never changes.

To sum up:
If your ROS MASTER is on the car (you started teleop on the car and have launched anything else) then
**On Car**
`ROS_MASTER_URI = http://localhost:11311`  
`ROS_IP = CAR_IP`

**On Laptop**  
`ROS_MASTER_URI = http://CAR_IP:11311`  
`ROS_IP = LAPTOP_IP`

If this is still confusing, please let me know and I'll rephrase!

**Comment by palanc** [comment]
Hi @Lilly-yang - In addition to @schmittlema 's followup questions, could you also let us know whether your laptop is running a virtual machine or has a native Ubuntu OS? If the laptop is running a virtual machine, you may need to change some network settings in order to connect to the robot's default access point..

**Comment by Lilly-yang** [comment]
I have solved it. It's just that I still use wifi 😂 . I tried to follow the [first step](https://mushr.io/tutorials/first_steps/) to set a fixed IP, but after completing all the steps, the robot can't automatically (or manually) connect to the wifi (or hotspot).

Thank you @schmittlema for the detailed explanation. Now I can connect the robot to rivz on my laptop properly. It's just that I may need to check the new IP when it changes.

Thank you @planc509 for the supplement. I am using the native Ubuntu OS.  It did take me some time to set up the proper environment but it was worth it.

Have a good weekend!

**Comment by Lilly-yang** [comment]
Hi, it's me again.
I'm still trying to get the hotspot issue sorted out. All this time I've been using IP share of my library's internet. But it is not static. So I have to correct the IP in ~/.bashrc often which requires me to connect the screen, mouse and keyboard...It's a pain.

I followed the tutorial to set up an access point on the Jetson Nano. Still the same problem: I can't connect to it on my laptop with "ssh robot@10.42.0.1" or "ssh robot@10.42.0.1 -X". But I can connect its wifi(Robot AP) with my phone and this can be assumed that the hotspot is working fine. Can you help me figure out why I'm having ssh connection problems? Thank you!

**Reply by schmittlema** [reply]
Hey Lilly!
So if I am understanding correctly, your laptop is connected to Robot AP and you are trying to ssh into 10.42.0.1?
on the robot could you paste the output of the following command
`ifconfig`

**Reply by Lilly-yang** [reply]
Yes, I checked it out.
![image](https://user-images.githubusercontent.com/39558023/160187624-874e5a6d-26fe-4af9-8e37-8bc048dc4f0d.png)
My laptop is not connected to the Robot AP. I use my smartphone to connect to confirm if the Robot AP is available. It turns out Robot AP can be connected. It just can't be `ssh`.

**Reply by schmittlema** [reply]
Ok so your laptop wifi needs to be on Robot AP (the robot's network) then you should be able to ssh

**Reply by Lilly-yang** [reply]
It works! I can't believe it's actually because I didn't connect to its wifi. I really didn't see this step in the tutorial. Thank you, I finally solved this problem that has been bothering me for a long time. Have a good weekend!

**Reply by schmittlema** [reply]
Sorry about that, we will update the tutorial to be more clear! Happy Friday!
